Approximate_convex_decomposition/include/CGAL/internal/Approximate_convex_decomposition/approx_decomposition.h

```diff
@@ -0,0 +1,429 @@
+        m_decimation_map = boost::get(decimation_props_t(), m_graph);
+
+        // add vertices
+        // fill up cluster properties (single face) for all vertices
+        BOOST_FOREACH(face_descriptor face, vertices(dual))
+        {
+            Cluster_properties props;
+            
+            props.concavity = 0;
+            props.faces.push_back(face);
+
+            BOOST_FOREACH(vertex_descriptor vert, vertices_around_face(halfedge(face, m_mesh), m_mesh))
+            {
+                props.conv_hull_pts.push_back(get(CGAL::vertex_point, m_mesh)[vert]);
+            }
+
+            props.bbox = CGAL::Polygon_mesh_processing::face_bbox(face, m_mesh);
+
+            face_graph_map[face] = add_vertex(props, m_graph);
+        }
+
+        // add edges
+        // decimation properties are filled up in the later calls to update_edge
+        BOOST_FOREACH(edge_descriptor edge, edges(dual))
+        {
+            Decimation_properties props;
+
+            add_edge(face_graph_map[source(edge, dual)], face_graph_map[target(edge, dual)], props, m_graph);
+        }
+    }
+
+    /**
+     * Constructs cluster of two smaller clusters and computes decimation cost.
+     */
+    void update_edge(graph_edge_descriptor edge, double concavity_threshold, double alpha_factor)
+    {
+        Decimation_properties& decimation_props = m_decimation_map[edge];
+
+        graph_vertex_descriptor vert_1 = source(edge, m_graph), vert_2 = target(edge, m_graph);
+
+        Cluster_properties& cluster_1_props = m_cluster_map[vert_1];
+        Cluster_properties& cluster_2_props = m_cluster_map[vert_2];
+
+        // convex hull points
+        std::vector<Point_3> common_hull_pts; // merged list of the lists of the convex hull points of two clusters
+        
+        CGAL_assertion(cluster_1_props.conv_hull_pts.size() >= 3);
+        CGAL_assertion(cluster_2_props.conv_hull_pts.size() >= 3);
+        
+        // add the convex hull points of the first cluster
+        BOOST_FOREACH(Point_3 p, cluster_1_props.conv_hull_pts)
+        {
+            common_hull_pts.push_back(p);
+        }
+        // add the convex hull points of the second cluster
+        BOOST_FOREACH(Point_3 p, cluster_2_props.conv_hull_pts)
+        {
+            common_hull_pts.push_back(p);
+        }
+
+        // compute convex hull on the merged list of points
+        TriangleMesh conv_hull;
+        CGAL::convex_hull_3(common_hull_pts.begin(), common_hull_pts.end(), conv_hull);
+
+        // fill up the list of the convex hull points of produced cluster after decimation
+        decimation_props.new_cluster_props.conv_hull_pts.clear();
+        BOOST_FOREACH(vertex_descriptor vert, vertices(conv_hull))
+        {
+            decimation_props.new_cluster_props.conv_hull_pts.push_back(get(CGAL::vertex_point, conv_hull)[vert]);
+        }
+        CGAL_warning(decimation_props.new_cluster_props.conv_hull_pts.size() > 3);
```

Comment:
what happen if you have some degenerate faces at initialization?

Reply:
Problems might occur with convex hull computation. Can we get rid of such faces in the beginning?

Reply:
if you have only 2 points the concavity measure is trivial